### PR TITLE
Feature: Support db get for duptables (CLI)

### LINKED ISSUE #2836: Add duptable support for `reth db get`
### Describe the feature

`reth db get` does not currently support duptables but it would be nice if it did

### Additional context

https://github.com/paradigmxyz/reth/pull/2772#pullrequestreview-1436354459

### PR DESCRIPTION
**Pull request by pyropy**
Continues upon #2897 . Closes #2836

### CODE REVIEW

**Comment by pyropy** [comment]
@gakonst here's a PR for the issue I've mentioned in #4282. Could we get issue #2836 assigned to me?

File: bin/reth/src/db/clear.rs [hunk 1]
```diff
@@ -38,4 +38,12 @@ impl<DB: Database> TableViewer<()> for ClearViewer<'_, DB> {
 
         Ok(())
     }
+
+    fn view_dupsort<T: Table>(&self) -> Result<(), Self::Error> {
```
Comment (onbjerg) [inline bin/reth/src/db/clear.rs:42]: can we make the bound `DupSort`?

File: bin/reth/src/db/get.rs [hunk 2]
```diff
@@ -38,6 +40,12 @@ impl Command {
 
         serde_json::from_str::<T::Key>(&self.key).map_err(|e| eyre::eyre!(e))
     }
+    /// Get an instance of subkey for given dupsort table
+    fn table_subkey<T: Table + DupSort>(&self) -> Result<T::SubKey, eyre::Error> {
```
Comment (onbjerg) [inline bin/reth/src/db/get.rs:44]: IIRC `Table` is implied for `DupSort`, `DupSort` should be fine?

File: bin/reth/src/db/get.rs [hunk 3]
```diff
@@ -63,8 +69,26 @@ impl<DB: Database> TableViewer<()> for GetValueViewer<'_, DB> {
 
         Ok(())
     }
-}
 
+    fn view_dupsort<T: Table + DupSort>(&self) -> Result<(), Self::Error> {
+        // error here
```
Comment (onbjerg) [inline bin/reth/src/db/get.rs:74]: leftover?

File: crates/storage/db/src/tables/mod.rs [hunk 4]
```diff
@@ -159,32 +162,32 @@ macro_rules! tables {
...
-    (StorageChangeSet, TableType::DupSort),
-    (HashedAccount, TableType::Table),
-    (HashedStorage, TableType::DupSort),
-    (AccountsTrie, TableType::Table),
-    (StoragesTrie, TableType::DupSort),
-    (TxSenders, TableType::Table),
-    (SyncStage, TableType::Table),
-    (SyncStageProgress, TableType::Table),
-    (PruneCheckpoints, TableType::Table)
+    (CanonicalHeaders, TableType::Table, view),
```
Comment (onbjerg) [inline crates/storage/db/src/tables/mod.rs:165]: guess this is ok, wish we didn't have to though

**Review by onbjerg — DISMISSED** [review]

File: crates/storage/db/src/tables/mod.rs (hunk 4 above)
Comment (pyropy) [reply crates/storage/db/src/tables/mod.rs:165]: Me too, I was trying to find a better way to refactor the macro. I am unhappy with this way but I lack ideas. I'm up for suggestions tho

**Comment by pyropy** [comment]
@onbjerg How could I test this? I am new to the project so I am unfamiliar with the workflows.

**Comment by dragan2234** [comment]
@pyropy find a table with duplicated keys and try to get the value with the subkey with your newly created command. I don't know if this is enough for testing(if you should write automated tests - that's the team decision)

**Comment by rkrasiuk** [comment]
@pyropy you can simply initialize a holesky node and query for the storage slots of the deposit contract. do you have availability to bring this PR to completion?

**Comment by pyropy** [comment]
> @pyropy you can simply initialize a holesky node and query for the storage slots of the deposit contract. do you have availability to bring this PR to completion?

Hey yes I'll do it over the weekend, I am in on-site camp with my company so I cannot do it ATM.

**Comment by rkrasiuk** [comment]
@pyropy fyi, the dupsort tables were wrongfully disabled on `db list` https://github.com/paradigmxyz/reth/pull/5068

**Comment by pyropy** [comment]
I've synced holesky network and tried out the cli, here's the example for `PlainStorageState` table:

```
./target/debug/reth db get PlainStorageState 0x4242424242424242424242424242424242424242 0x0000000000000000000000000000000000000000000000000000000000000004 --chain=holesky
{
  "key": "0x0000000000000000000000000000000000000000000000000000000000000004",
  "value": "0x156663e7936e7dd38052dbffe5c8a41e507ccef69867efd1af269df0d0c7a8fe"
}
```

**Comment by rkrasiuk** [comment]
@pyropy sg! my remaining concern is the need for `view_dupsort` function. do we still need to differentiate given #5068?

**Comment by pyropy** [comment]
> @pyropy sg! my remaining concern is the need for `view_dupsort` function. do we still need to differentiate given #5068?

Good call, let me take a look at that today and refactor my PR if needed.

**Comment by gakonst** [comment]
Hi @pyropy wonder if there's any next steps for this PR? Thanks for taking the time!

**Comment by pyropy** [comment]
> Hi @pyropy wonder if there's any next steps for this PR? Thanks for taking the time!

Hey I think it should be good. If you can give it a review you I'd appreciate it 🙏🏻

File: bin/reth/src/db/tui.rs [hunk 5]
```diff
@@ -195,7 +195,7 @@ where
     /// Fetch the next page of items
     fn next_page(&mut self) {
         if self.skip + self.count >= self.total_entries {
-            return
+            return;
```
Comment (rkrasiuk) [inline bin/reth/src/db/tui.rs:198]:
```suggestion
            return
```

File: bin/reth/src/db/tui.rs [hunk 6]
```diff
@@ -205,7 +205,7 @@ where
     /// Fetch the previous page of items
     fn previous_page(&mut self) {
         if self.skip == 0 {
-            return
+            return;
```
Comment (rkrasiuk) [inline bin/reth/src/db/tui.rs:208]:
```suggestion
            return
```

File: bin/reth/src/db/tui.rs [hunk 7]
```diff
@@ -311,7 +311,7 @@ where
             }
         }
 
-        return Ok(false)
+        return Ok(false);
```
Comment (rkrasiuk) [inline bin/reth/src/db/tui.rs:314]:
```suggestion
        return Ok(false)
```

File: bin/reth/src/utils.rs [hunk 8]
```diff
@@ -153,16 +153,16 @@ impl<'a, DB: Database> DbTool<'a, DB> {
...
                     match &*bmb {
                         Some(searcher) => {
-                            if searcher.find_first_in(&value).is_some() ||
-                                searcher.find_first_in(&key).is_some()
+                            if searcher.find_first_in(&value).is_some()
+                                || searcher.find_first_in(&key).is_some()
                             {
                                 hits += 1;
-                                return result()
+                                return result();
```
Comment (rkrasiuk) [inline bin/reth/src/utils.rs:160]:
```suggestion
                                return result()
```

File: bin/reth/src/utils.rs [hunk 9]
```diff
@@ -153,16 +153,16 @@ impl<'a, DB: Database> DbTool<'a, DB> {
...
                             }
                         }
                         None => {
                             hits += 1;
-                            return result()
+                            return result();
```
Comment (rkrasiuk) [inline bin/reth/src/utils.rs:165]:
```suggestion
                            return result()
```

File: bin/reth/src/db/list.rs [hunk 10]
```diff
@@ -130,4 +134,45 @@ impl TableViewer<()> for ListTableViewer<'_> {
...
+            let total_entries = stats.entries();
+            if self.args.skip > total_entries - 1 {
+                error!(
+                    target: "reth::cli",
+                    "Start index {start} is greater than the final entry index ({final_entry_idx}) in the table {table}",
+                    start = self.args.skip,
+                    final_entry_idx = total_entries - 1,
+                    table = self.args.table.name()
+                );
+                return Ok(());
```
Comment (rkrasiuk) [inline bin/reth/src/db/list.rs:151]:
```suggestion
                return Ok(())
```

File: bin/reth/src/db/list.rs [hunk 11]
```diff
@@ -61,7 +65,7 @@ impl Command {
             .as_ref()
             .map(|search| {
                 if let Some(search) = search.strip_prefix("0x") {
-                    return hex::decode(search).unwrap()
+                    return hex::decode(search).unwrap();
```
Comment (rkrasiuk) [inline bin/reth/src/db/list.rs:68]:
```suggestion
                    return hex::decode(search).unwrap()
```

File: bin/reth/src/utils.rs [hunk 12]
```diff
@@ -143,7 +143,7 @@ impl<'a, DB: Database> DbTool<'a, DB> {
 
                     let result = || {
                         if filter.only_count {
-                            return None
+                            return None;
```
Comment (rkrasiuk) [inline bin/reth/src/utils.rs:146]:
```suggestion
                            return None
```

**Comment by rkrasiuk** [comment]
@pyropy thanks for the PR, we'll be merging it shortly. refactored the tables macro to a bit more amicable format

**Comment by pyropy** [comment]
> @pyropy thanks for the PR, we'll be merging it shortly. refactored the tables macro to a bit more amicable format

Seen your refactor, good job. This is a really good way for me to learn how to do rust properly as I mainly write golang.

File: bin/reth/src/db/get.rs [hunk 13]
```diff
@@ -74,8 +81,25 @@ impl<DB: Database> TableViewer<()> for GetValueViewer<'_, DB> {
...
+                println!("{}", serde_json::to_string_pretty(&content)?);
+            }
+            None => {
+                error!(target: "reth::cli", "No content for the given table subkey.");
+            }
+        };
+        Ok(())
+    }
+}
 /// Map the user input value to json
```
Comment (DaniPopes) [inline bin/reth/src/db/get.rs:103]:
```suggestion
}

/// Map the user input value to json
```

File: bin/reth/src/db/get.rs [hunk 14]
```diff
@@ -1,7 +1,11 @@
 use crate::utils::DbTool;
 use clap::Parser;
```
Comment (DaniPopes) [inline bin/reth/src/db/get.rs:3]:
```suggestion
```

File: bin/reth/src/db/get.rs [hunk 15]
```diff
@@ -41,6 +43,12 @@ impl Command {
 
         serde_json::from_str::<T::Key>(&self.key).map_err(|e| eyre::eyre!(e))
     }
+    /// Get an instance of subkey for given dupsort table
```
Comment (DaniPopes) [inline bin/reth/src/db/get.rs:46]:
```suggestion
    }

    /// Get an instance of subkey for given dupsort table
```

**Review by DaniPopes — APPROVED** [review]
LGTM. Fmt nits. IMO the two table macros should be merged in a follow-up.

**Pull request merged**